common/persistence/sql/sqldriver/sharded.go

```diff
@@ -0,0 +1,195 @@
+
+	"github.com/uber/cadence/common/persistence/sql/sqlplugin"
+)
+
+type (
+	// sharded is the driver querying a group of SQL databases as sharded solution
+	sharded struct {
+		dbs           []*sqlx.DB // this is for starting a transaction, or executing any non transaction query
+		tx            *sqlx.Tx   // this is a reference of a started transaction
+		useTx         bool       // if tx is not nil, the methods from commonOfDbAndTx should use tx
+		currTxShardID int        // which shard is current tx started from
+	}
+
+	shardedSqlExecResult struct{}
+)
+
+// newShardedSQLDriver returns a driver querying a group of SQL databases as sharded solution.
+// xdbs is the list of connections to the sql instances. The length of the list of the list is the totalNumShards
+// dbShardID is needed when tx is not nil. It means a started transaction in the shard.
+func newShardedSQLDriver(xdbs []*sqlx.DB, xtx *sqlx.Tx, dbShardID int) Driver {
+	driver := &sharded{
+		dbs: xdbs,
+		tx:  xtx,
+	}
+	if xtx != nil {
+		driver.useTx = true
+		driver.currTxShardID = dbShardID
+	}
+	return driver
+}
+
+// below are shared by transactional and non-transactional, if s.tx is not nil then use s.tx, otherwise use s.db
+
+func (s *sharded) ExecContext(ctx context.Context, dbShardID int, query string, args ...interface{}) (sql.Result, error) {
+	if dbShardID == sqlplugin.DbShardUndefined || dbShardID == sqlplugin.DbAllShards {
+		return nil, fmt.Errorf("invalid dbShardID %v shouldn't be used to ExecContext, there must be a bug", dbShardID)
+	}
+	if s.useTx {
+		if s.currTxShardID != dbShardID {
+			return nil, fmt.Errorf("dbShardID %v doesn't match with started transaction %v, must be a bug", dbShardID, s.currTxShardID)
+		}
+		return s.tx.ExecContext(ctx, query, args...)
+	}
+	return s.dbs[dbShardID].ExecContext(ctx, query, args...)
+}
+
+func (s *sharded) NamedExecContext(ctx context.Context, dbShardID int, query string, arg interface{}) (sql.Result, error) {
+	if dbShardID == sqlplugin.DbShardUndefined || dbShardID == sqlplugin.DbAllShards {
+		return nil, fmt.Errorf("invalid dbShardID %v shouldn't be used to NamedExecContext, there must be a bug", dbShardID)
+	}
+	if s.useTx {
+		if s.currTxShardID != dbShardID {
+			return nil, fmt.Errorf("dbShardID %v doesn't match with started transaction %v, must be a bug", dbShardID, s.currTxShardID)
+		}
+		return s.tx.NamedExecContext(ctx, query, arg)
+	}
+	return s.dbs[dbShardID].NamedExecContext(ctx, query, arg)
+}
+
+func (s *sharded) GetContext(ctx context.Context, dbShardID int, dest interface{}, query string, args ...interface{}) error {
+	if dbShardID == sqlplugin.DbShardUndefined || dbShardID == sqlplugin.DbAllShards {
+		return fmt.Errorf("invalid dbShardID %v shouldn't be used to GetContext, there must be a bug", dbShardID)
+	}
+	if s.useTx {
+		if s.currTxShardID != dbShardID {
+			return fmt.Errorf("dbShardID %v doesn't match with started transaction %v, must be a bug", dbShardID, s.currTxShardID)
```

Comment:
nit: this error is repeated everywhere, worth creating a help method

Reply:
Good idea. Will do.